pkg/vsphere/tasks/waiter.go

```diff
@@ -27,12 +28,47 @@ import (
 	"github.com/vmware/govmomi/vim25/progress"
 	"github.com/vmware/govmomi/vim25/soap"
 	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/pkg/trace"
 )
 
 const (
 	maxBackoffFactor = int64(16)
+
+	VMContextObjectKey = "VMContextObject"
+)
+
+var (
+	m             sync.RWMutex
+	errorHandlers []ErrorHandler
 )
 
+type ErrorHandler interface {
+	// Handle returns true if the error is expected. Otherwise, return false. If any error happens during error handling, return (true, error).
+	Handle(ctx context.Context, err error) (bool, error)
+}
+
+type TaskInProgressHandler struct{}
+
+func (h *TaskInProgressHandler) Handle(ctx context.Context, err error) (bool, error) {
+	if isTaskInProgress(err) {
+		// TaskInProgress error, no need to fail here
+		log.Debugf("TaskInProgress error, continue")
+		return true, nil
+	}
+	return false, nil
+}
+
+func init() {
+	RegisterErrorHandler(&TaskInProgressHandler{})
+}
+
+func RegisterErrorHandler(handler ErrorHandler) {
```

Comment:
hmm... We need different logic for different errors. For this invalid state errors, one additional parameter definitely works. But for a common function, signature change means lots of changes unrelated.
And in the future we'll find more errors, and more error handling logic from anywhere might be required.
This separate error handling registration will be a more flexible solution compared to the signature change, I think.

Taking about the multiple handlers and execution sequence, I think this is defined for error handling only. Which means each time, there is one definitive error happens, and suppose only one corresponding error handler should be executed. Every other handlers will figure out that's not the one they can help, and return false without error. Which means next error handler will be tried.
So the execution sequence should not impact the result.